### ISSUE TITLE
Custom enumerators

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#92]
**Is your feature request related to a problem?**
Implementing custom enumerators in VB6 is difficult. twinBASIC could make this easier with some language-level support.

**Describe the solution you'd like**
Built-in interfaces (`IEnumerable` and `IEnumerable(Of T)`), which would contain a `NewEnum` function (which would be automatically assigned Procedure ID = -4) which returns a type which gets special-cased by the compiler to support [IEnumVARIANT](https://docs.microsoft.com/en-us/windows/win32/api/oaidl/nn-oaidl-ienumvariant).

**Describe alternatives you've considered**
Alternately, special keywords could be used, not sure what that would be though.

**Additional context**
The `For Each` construct can be supported in a custom collection class by adding an `IUnknown` function called `NewEnum` and setting it's procedure Id to -4:

```
' @ProcedureId(-4)
Public Function NewEnum As IUnknown
  Set NewEnum = _collection.[_NewEnum]
End Function
```
The function must return a pointer to an `IEnumVARIANT` variant. And therein lies the problem - this interface is not implementable in VB6. [There is a workaround](https://www.mvps.org/vbvision/Super_Collections.htm), but it requires several hacks and some custom IDL.

It would be great if we could do this:

```
Public Class List(Of T)
  Dim _enumerator As Enumerator(Of T)

  Public Sub New()
     _enumerator = new Enumerator(Of T)(Me)
  End Sub

  Public Function NewEnum As IEnumerator(Of T) ' Automatically assigned Procedure ID -4
    Set NewEnum = _enumerator
  End Function
End Class

Public Class Enumerator(Of T)
  Implements IEnumerator(Of T) ' <------ Compiles to IEnumVariant

  ' Implementation skipped
End Class
```

**Comment by bclothier** [comment]
Just as a note... `IEnumVARIANT` can only return a `Variant`. There are several articles out there explaining how to create a custom collection class but ultimately it's possible to write a VB* code like this that compiles but will blow up runtime:

```
Dim notThatThing As NotThatThing

For Each notThatThing In ACollectionOfThings
Next
```

The non-generic `IEnumerator` should compile down to  `IEnumVARIANT`, but the generic version will have to be a private implementation detail within tB code.

**Comment by mansellan** [comment]
Ok cool, makes sense. Wasn't sure how it would play with tB generics.

**Comment by FullValueRider** [comment]
It might make more sense to allow initialisers for collections.  If it were possible to pass arrays to a collection (one for the keys and one for the values) you getalmost everything @Mansellan has requested as well as making the collection object a much more usable object.

e.g.

Dim myColl as collection = new collection(array1,array2)

**Comment by vbRichClient** [comment]
In VB6 (and VBA in 32Bit-Office) we can already do this (directly in a Form-Class):
```
Option Explicit

Implements vbIEnumerable

Private Keys(1 To 3) As String, Vals(1 To 3) As Double, Member

Private Sub Form_Load()
  Keys(1) = "Prop1": Vals(1) = 1.1
  Keys(2) = "Prop2": Vals(2) = 2.2
  Keys(3) = "Prop3": Vals(3) = 3.3
  
  With New_c.ArrayList(vbString)
    For Each Member In New_c.VBI.EnumerateOn(Me, LBound(Keys))
       .Push Member
    Next
    Debug.Print "[" & .Join(", ") & "]" 'finally join the pushed JsonObj-parts into a JSON-array
  End With
End Sub
 
Private Function vbIEnumerable_ElementForIndex(ByVal Index As Long, UserData, CancelEnumeration As Boolean)
  If Index > UBound(Keys) Then CancelEnumeration = True: Exit Function
  
  'return the members of 2 different arrays as a properly encoded JSON-Object-String
  vbIEnumerable_ElementForIndex = "{""" & Keys(Index) & """:" & Str(Vals(Index)) & "}"
End Function
```
Which then prints out the following, properly encoded JSON-array-string: 
`[{"Prop1": 1.1}, {"Prop2": 2.2}, {"Prop3": 3.3}]`

The only requirement for the above snippet to work, is a reference to vbRichClient5 or (the newer) RC6.

I guess what I'm trying to say is, that "pure" IEnumVariant-based Enumerations -
only have to be properly supported at compiler-level (which twinBasic currently doesn't).

Advanced flexibility with Enumerations, can then be implemented at runtime-lib-level -or as in case of the referenced RC6-lib, 
at the level of any Helper-Lib which returns proper IEnumVariant-wrapping Objects (behind the In-Keyword of a For Each Loop).

Olaf

**Comment by Kr00l** [comment]
There are plenty of ways to achieve this in VB6. IMO a good way is to use [lightweight-COM described by DEXWERX](https://www.vbforums.com/showthread.php?854963-VB6-IEnumVARIANT-For-Each-support-without-a-typelib).
So I think there is no hurry for tB to have a "native" solution and first should work (at least first) with the current solutions in place.

**Comment by Greedquest** [comment]
@Kr00l I think language support is probably quite key (I know you aren't suggesting it isn't) but imo that com workaround is too buggy and slow for me to use even in proof of concept code, the unmanaged memory just risks crashes while debugging.

Also does `IEnumerable` have lazy evaluation built in - you provide either an array of values or a generator function, or should I put in a separate feature request for something like a `yield` keyword to allow this?

**Comment by mansellan** [comment]
@Greedquest yes, `IEnumerable` has callbacks for `MoveNext` and `Current`, so can be evaluated lazily. That said, a `Yield Return` syntax would allow simpler implementation, so would be a nice-to-have IMO.

**Comment by WaynePhillipsEA** [comment]
I think this is a great idea.  Tomorrow we'll have IDispatch support in tB classes, so we can then use the legacy DISPID_ENUM(-4) method to supply the enumerator.  But going forward, I'll be adding this native version, or something very similar.

**Comment by mansellan** [comment]
[As bclothier said](https://github.com/WaynePhillipsEA/twinbasic/issues/92#issuecomment-826361804), it makes more sense to base the `IEnumerable` (non-generic) interface on the [IEnumVARIANT interface](https://docs.microsoft.com/en-us/windows/win32/api/oaidl/nn-oaidl-ienumvariant) than on the .Net equivalent.

The `IEnumVARIANT` interface inherits `IUnknown`, and has the following members:

```c
HRESULT Clone(IEnumVARIANT **ppEnum);
HRESULT Next(ULONG  celt, VARIANT *rgVar,  ULONG  *pCeltFetched);
HRESULT Reset();
HRESULT Skip(ULONG celt);
```

Note that these use an unsigned long, which is not yet supported in twinBASIC (seee #81).

If we could have an `IEnumerable` interface of the following:

```vb
Interface IEnumerable
  Public Function Clone(ByRef elements As IEnumerable) As Boolean
  Public Function Next(ByVal elementCount As ULong, ByRef elements As Variant(), ByRef returnedElementCount As ULong) As Boolean
  Public Function Reset() As Boolean
  Public Function Skip(ByVal elementCount As ULong) As Boolean
End Interface
```

Could twinBASIC handle the conversion to / from `IEnumVariant`? Ideally, `IEnumerable` and `IEnumVARIANT` should be freely interchangeable. 

Note, I suspect that `Clone` and `Skip` are not used by the VB `For Each` construct, however having them available on the interface is probably best for interop.

**Comment by mansellan** [comment]
Also, I assumed the interface needed `Boolean` functions to handle the `HRESULT`, but actually maybe they are better as Subs?

**Comment by WaynePhillipsEA** [comment]
@mansellan The VB.NET IEnumerator internally implements IEnumVARIANT, and is simpler, having only 3 methods (prop-get Current / MoveNext / Reset). 

I suggest that twinBASIC simply offers the same/equivalent IEnumerator interface, and handles the IEnumVARIANT for you behind the scenes. 

Also, for a fully conformant implementation, the IEnumVARIANT::Next method requires support for returning a non-error HRESULT of S_FALSE (value 1), which is something that you can't ordinarily do in VBx or tB, so I do think the raw implementation of IEnumVARIANT is best handled by the compiler.

**Comment by WaynePhillipsEA** [comment]
> Also, I assumed the interface needed `Boolean` functions to handle the `HRESULT`, but actually maybe they are better as Subs?

Class members in tB (and VBx) return HRESULTs for you, based on the local error handling.  A return value defined on a class member is  actually passed as an out-reference parameter, as the final parameter.

VB.NET gives more control over HRESULT handling, using the `[PreserveSig]` attribute, but we haven't yet got anything similar in tB.

**Comment by Kr00l** [comment]
> Also, for a fully conformant implementation, the IEnumVARIANT::Next method requires support for returning a non-error HRESULT of S_FALSE (value 1), which is something that you can't ordinarily do in VBx or tB, so I do think the raw implementation of IEnumVARIANT is best handled by the compiler.

I know that lightweight COM can circumvent that HRESULT "problem" by just having a Function which returns Long. It's not a big deal to still do that.

However, would it be technically possible to somehow allow a return value for a "Sub" (=HRESULT) ? I can imagine that COM error-ing stuff could dissalow it..
It may be convenient here and there when dealing with type libraries..

**Comment by mansellan** [comment]
@Kr00l Reading up on `[PreserveSig]`, it seems to do exactly that. I hadn't heard of it before Wayne mentioned it.

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.preservesigattribute?view=net-5.0

**Comment by FullValueRider** [comment]
Has twinBasic achieved this goal yet, or is it still on the 'to do' list

**Comment by WaynePhillipsEA** [comment]
I'll look at this very soon

**Comment by wqweto** [comment]
Btw, VBx uses `IEnumVARIANT::Next` method only and in a limited fashion only (requesting a single element in `celt`).

Probably not a good idea to base TB's enumerators imp on the complete COM interface as the rest of the methods of `IEnumVARIANT` are dead cruft already.

**Comment by Kr00l** [comment]
Any chance to get a safe handler for `IEnumString` as well? It's also useful for some API's. (e.g. the SHAutocomplete API)

So, for example:
`Implements IEnumerator(Of String)` ' <- IEnumString
`Implements IEnumerator(Of Variant)` ' <- IEnumVARIANT
something else would cause a compile error.

**Comment by bclothier** [comment]
Apologies for piling on but I want to point out that one scenario we should consider is supporting enumerable collection for `WithEvents`. We currently have the support for using an array which is great when the members are known at the compile time but I think there are also scenarios where members are added/removed at runtime and therefore an array would be quite expensive. Thus, it would be nice to have an enumerator that can be used with the `WithEvents` to forward the events from the members.

**Comment by mansellan** [comment]
> something else would cause a compile error

I respectfully disagree. `IEnumerable (Of Integer)` should be compilable in twinBASIC, and templated for enumerating through Integers from twnBASIC callers. At the COM boundary, it should be expressed as `IEnumVARIANT`, and possibly runtime checked for a subtype of `Ineger`.

**Comment by sancarn** [comment]
> Any chance to get a safe handler for `IEnumString`

Surely it'd be better if Variant's could just be a generic type. I.E. 

```vb
Interface IEnumVARIANT<T>
  Public Function Clone(ByRef elements As IEnumerable) As Boolean
  Public Function Next(ByVal elementCount As ULong, ByRef elements As Variant<T>, ByRef returnedElementCount As ULong) As Boolean
  Public Function Reset() As Boolean
  Public Function Skip(ByVal elementCount As ULong) As Boolean
End Interface
```

At that point you could Identify the type of variants at compile time, helping massively with intellisense.

**Comment by bclothier** [comment]
I see what I did wrong. I shouldn't have called it `TestCustomEnumeartor` (sic). I should have called it `LameHalfDonkeyedCustomEnumeratorOnlyGoodForDemostratingOneLevelOfEnumeration`. 😄 

That said, keep in mind that `IEnumVARIANT::Clone` is supposed to return a copy with the same state, so a full implementation would be more complicated as you'd need to share the state between the collection and the implementation of the `IEnumVARIANT`.  Again, I don't know if anyone calls `IEnumVARIANT::Clone` but it's documented to return with the same state, rather than a reset state as in your example.
